File: src/main/java/eu/rossng/camle/Camle.java
package eu.rossng.camle;

import org.antlr.v4.runtime.ANTLRFileStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.tree.ParseTree;

import java.lang.reflect.Array;

/**
 * A compiler for the Whilst language, built for COMS22201 at the University of Bristol.
 */
public class Camle {
    public static void main(String[] args) {
        System.out.println("\n=============================================================");
        System.out.println("CAMLE - Compiler to Abstract Machine for Language Engineering");
        System.out.println("=============================================================\n");
        String opt = "", inFile = "", outFile = "";
        int pos;
        if (Array.getLength(args) == 1 && args[0].charAt(0) != '-') {
            opt = "";
            inFile = args[0];
        } else if (Array.getLength(args) == 2 && args[0].charAt(0) == '-' &&
                args[1].charAt(0) != '-') {
            opt = args[0];
            inFile = args[1];
        } else {
            System.out.println("Usage: <jar> <option> filename");
            System.out.println();
            System.out.println("Where <jar> executes the CAMLE jar, for example:");
            System.out.println(" java -jar target/camle-maven-1.0-SNAPSHOT-jar-with-dependencies.jar");
            System.out.println();
            System.out.println("And where <option> is one of:\n -lex, -syn, -irt");
            System.out.println();
            System.out.println("And where filename is the name of a file containing Whilst source code");
            System.exit(1);
        }

        try {
            ANTLRFileStream input = new ANTLRFileStream(inFile);
            SynLexer lexer = new SynLexer(input);
            CommonTokenStream tokens = new CommonTokenStream(lexer);
            SynParser parser = new SynParser(tokens);
            ParseTree tree = parser.program();

            switch (opt) {
                case "-lex":
                    Token T;
                    T = lexer.nextToken();
                    while (T.getType() != -1) {
                        System.out.println(T.getType() + " \"" + T.getText() + "\"");
                        T = lexer.nextToken();
                    }
                    System.exit(0);
                case "-syn":
                    System.out.println(tree.toStringTree(parser));
                    System.exit(0);
                default:
                    System.out.println("Support for -irt has not yet been implemented.");
                    System.exit(0);
            }
        } catch (Exception e) {
            System.out.print(e.getLocalizedMessage());
            System.exit(1);
        }
    }
}

File: src/main/kotlin/eu/rossng/camle/Camle.kt
package eu.rossng.camle

import org.antlr.v4.runtime.ANTLRFileStream
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.Token
import java.io.FileOutputStream
import java.io.OutputStream
import java.io.PrintStream

/**
 * A compiler for the Whilst language, built for COMS22201 at the University of Bristol.
 */
fun main(args: Array<String>) {
    println("\n=============================================================")
    println("CAMLE - Compiler to Abstract Machine for Language Engineering")
    println("=============================================================\n")
    var opt = ""
    var inFile = ""
    val outFile = ""
    val pos: Int
    if (args.size == 1 && args[0].get(0) != '-') {
        opt = ""
        inFile = args[0]
    } else if (args.size == 2 && args[0].get(0) == '-' && args[1].get(0) != '-') {
        opt = args[0]
        inFile = args[1]
    } else {
        println("Usage: <jar> <option> filename")
        println()
        println("Where <jar> executes the CAMLE jar, for example:")
        println(" java -jar target/antlr4-camle-jar-with-dependencies.jar")
        println()
        println("And where <option> is one of:\n -lex, -syn, -irt, -cg")
        println()
        println("And where filename is the name of a file containing Whilst source code")
        System.exit(1)
    }

    try {
        val input = ANTLRFileStream(inFile)
        val lexer = WhilstLexer(input)
        val tokens = CommonTokenStream(lexer)
        val parser = WhilstParser(tokens)

        when (opt) {
            "-lex" -> {
                var T: Token
                T = lexer.nextToken()
                while (T.type != -1) {
                    println("${T.type} \"${T.text}\"")
                    T = lexer.nextToken()
                }
                System.exit(0)
            }
            "-syn" -> {
                val tree = parser.program()
                println(tree.toStringTree(parser))
                System.exit(0)
            }
            "-irt" -> {
                val tree = parser.program()
                val whilstVisitor = WhilstFileVisitor()
                val irTree = whilstVisitor.visit(tree)
                System.out.println(irTree)
                System.exit(0)
            }
            "-cg" -> {
                val tree = parser.program()
                val whilstVisitor = WhilstFileVisitor()
                val irTree = whilstVisitor.visit(tree)
                val memory = Memory()
                val outputStream = PrintStream(FileOutputStream("out.ass"))
                val irVisitor = JouetteAsmGenerator(outputStream, memory)
                irVisitor.visit(irTree)
                outputStream.close()
            }
            else -> {
                println("Please provide valid arguments: -lex, -syn or -irt.")
                System.exit(1)
            }
        }
    } catch (e: Exception) {
        print(e.message)
        System.exit(1)
    }

}
